gitlab/resource_gitlab_managed_license.go

```diff
@@ -0,0 +1,165 @@
+		DeleteContext: resourceGitlabManagedLicenseDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"project": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "The ID of the project under which the managed license will be created.",
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				// GitLab's edit API doesn't allow us to edit the name, only
+				// the approval status.
+				ForceNew:    true,
+				Description: "The name of the managed license (I.e., 'Apache License 2.0' or 'MIT license')",
+			},
+			"approval_status": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     false,
+				ValidateFunc: validation.StringInSlice([]string{"approved", "blacklisted"}, true),
+				Description:  "Whether the license is approved or not. Only 'approved' or 'blacklisted' allowed.",
+			},
+		},
+	}
+}
+
+func resourceGitlabManagedLicenseCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+	project := d.Get("project").(string)
+
+	options := &gitlab.AddManagedLicenseOptions{
+		Name:           gitlab.String(d.Get("name").(string)),
+		ApprovalStatus: stringToApprovalStatus(d.Get("approval_status").(string)),
+	}
+
+	log.Printf("[DEBUG] create gitlab Managed License on Project %s, with Name %s", project, *options.Name)
+
+	addManagedLicense, _, err := client.ManagedLicenses.AddManagedLicense(project, options, gitlab.WithContext(ctx))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	licenseId := strconv.Itoa(addManagedLicense.ID)
+	d.SetId(buildTwoPartID(&project, &licenseId))
+
+	return resourceGitlabManagedLicenseRead(ctx, d, meta)
+}
+
+func resourceGitlabManagedLicenseDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+	project, licenseId, err := projectIdAndLicenseIdFromId(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	log.Printf("[DEBUG] Delete gitlab Managed License %s", d.Id())
+	_, err = client.ManagedLicenses.DeleteManagedLicense(project, licenseId, gitlab.WithContext(ctx))
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("failed to delete managed license %d for projec %s: %w", licenseId, project, err))
+	}
+
+	return nil
+}
+
+func resourceGitlabManagedLicenseRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+	project, licenseId, err := projectIdAndLicenseIdFromId(d.Id())
+	if err != nil {
+		diag.FromErr(err)
+	}
+
+	log.Printf("[DEBUG] read gitlab Managed License for project/licenseId %s/%d", project, licenseId)
+	license, _, err := client.ManagedLicenses.GetManagedLicense(project, licenseId, gitlab.WithContext(ctx))
+	if err != nil {
+		return diag.FromErr(err)
```

Comment:
Convention is to remove a resource from state if it no longer exists upstream.

```suggestion
		if is404(err) {
			log.Printf("[DEBUG] Managed License %s:%d no longer exists and is being removed from state", project, licenseId)
			d.SetId("")
			return nil
		}
		return diag.FromErr(err)
```